File: external/HDF-EOS/hdfeos/samples/definelevels.f
c
c	In this example we will (1) open the "PointFile" HDF file,
c	(2) attach to the three point structures, and (3) define
c	the fields within each point.
c
	program definelevels

	integer         status
	integer*4       ptfid, ptid
	integer*4       fieldtype(8), fieldorder(8)
	character*255   fldlist

	integer*4       ptopen, ptattach
	integer         ptdeflev, ptdeflink, ptdetach, ptclose
	
	integer  DFACC_RDWR
	parameter (DFACC_RDWR=3)

	integer  DFNT_CHAR8, DFNT_INT16, DFNT_INT32
	parameter (DFNT_CHAR8=4)
	parameter (DFNT_INT16=22)
	parameter (DFNT_INT32=24)
	
	integer  DFNT_FLOAT32, DFNT_FLOAT64
	parameter (DFNT_FLOAT32=5)
	parameter (DFNT_FLOAT64=6)
	
c
c	We first open the HDF point file, "PointFile.hdf".  Because this
c	file already exist and we wish to write to it, we use the
c	DFACC_RDWR access code in the open statement.  The ptopen
c	routine returns the point fileid, ptfid, which is used to
c	identify the file in subsequent routines.

	ptfid = ptopen("PointFile.hdf", DFACC_RDWR)


c
c	We next attach to the three point structures within the file
c       using the ptattach routine, identifying each point by its
c       name defined by the ptcreate routine previously.
c

	if (ptfid .ne. -1) then
	   
c
c      Define "Simple" point
c
	   
	   ptid = ptattach(ptfid, "Simple Point")
	   fldlist = "Time,Concentration,Species"


	   fieldtype(1) = DFNT_FLOAT64
	   fieldtype(2) = DFNT_FLOAT32
	   fieldtype(3) = DFNT_CHAR8
	   
	   fieldorder(1) = 1
	   fieldorder(2) = 4
	   fieldorder(3) = 4
	
	   status = ptdeflev(ptid, "Sensor", fldlist, fieldtype,
     1                       fieldorder)
	
	   status = ptdetach(ptid)
	   goto 5000
c
c      Define Fixed Buoy Point 
c	      

	   ptid = ptattach(ptfid, "FixedBuoy Point")


c     Define Description/Location Level

	   fldlist = "Label,Longitude,Latitude,DeployDate,ID"
	
	   fieldtype(1) = DFNT_CHAR8
	   fieldtype(2) = DFNT_FLOAT64
	   fieldtype(3) = DFNT_FLOAT64
	   fieldtype(4) = DFNT_INT32
	   fieldtype(5) = DFNT_CHAR8
	   
c   Note: order = 0 same as order = 1 for numeric scalars

	   fieldorder(1) = 8
	   fieldorder(2) = 0
	   fieldorder(3) = 0
	   fieldorder(4) = 0
	   fieldorder(5) = 1

	   status = ptdeflev(ptid, "Desc-Loc", fldlist, fieldtype, 
     1                       fieldorder)


c     Define Data Level

	   fldlist = "Time,Rainfall,Temperature,ID"
	
	   fieldtype(1) = DFNT_FLOAT64
	   fieldtype(2) = DFNT_FLOAT32
	   fieldtype(3) = DFNT_FLOAT32
	   fieldtype(4) = DFNT_CHAR8
	   
	   fieldorder(1) = 0
	   fieldorder(2) = 0
	   fieldorder(3) = 0
	   fieldorder(4) = 1

	   status = ptdeflev(ptid, "Observations", fldlist, 
     1			     fieldtype, fieldorder)

	   status = ptdeflink(ptid, "Desc-Loc", "Observations", "ID")

	   status = ptdetach(ptid)



c
c
c	Floating Buoy Point
c
	   ptid = ptattach(ptfid, "FloatBuoy Point")


c      Define Description Level */

	   fldlist = "Label,DeployDate,Weight,ID"
	
	   fieldtype(1) = DFNT_CHAR8
	   fieldtype(2) = DFNT_INT32
	   fieldtype(3) = DFNT_INT16
	   fieldtype(4) = DFNT_CHAR8
	   
	   fieldorder(1) = 8
	   fieldorder(2) = 0
	   fieldorder(3) = 0
	   fieldorder(4) = 1


	   status = ptdeflev(ptid, "Description", fldlist, 
     1                       fieldtype, fieldorder)


c	Define Data Level

	   fldlist = "Time,Longitude,Latitude,Rainfall,Temperature,ID"

	   fieldtype(1) = DFNT_FLOAT64
	   fieldtype(2) = DFNT_FLOAT64
	   fieldtype(3) = DFNT_FLOAT64
	   fieldtype(4) = DFNT_FLOAT32
	   fieldtype(5) = DFNT_FLOAT32
	   fieldtype(6) = DFNT_CHAR8
	   
	   fieldorder(1) = 0
	   fieldorder(2) = 0
	   fieldorder(3) = 0
	   fieldorder(4) = 0
	   fieldorder(5) = 0
	   fieldorder(6) = 1	

	   status = ptdeflev(ptid, "Measurements", fldlist, 
     1                       fieldtype, fieldorder)

	   status = ptdeflink(ptid, "Description", "Measurements", "ID")

	   status = ptdetach(ptid)
	
 5000	   continue
c
c      Close HDF file
c	
	   status = ptclose(swfid)

	   write(*,*) 'status close ', status
	endif
		   
	stop
	end

File: external/HDF-EOS/hdfeos/samples/DefineLevels.c
#include "mfhdf.h"

/* DefineLevels */

/*
 * In this example we will open the "PointFile" HDF file and define the point
 * levels.
 */


main()
{

    intn            status;
    int32           ptfid, PTid, fieldtype[8], fieldorder[8];
    char            fldlist[255];


    /*
     * We first open the HDF swath file, "PointFile.hdf".  Because this file
     * already exist and we wish to write to it, we use the DFACC_RDWR access
     * code in the open statement.  The PTopen routine returns the point file
     * id, ptfid, which is used to identify the file in subsequent routines.
     */

    ptfid = PTopen("PointFile.hdf", DFACC_RDWR);



    if (ptfid != -1)
    {


	/* Simple Point */
	/* ------------ */
	PTid = PTattach(ptfid, "Simple Point");
	strcpy(fldlist, "Time,Concentration,Species");

	fieldtype[0] = DFNT_FLOAT64;
	fieldtype[1] = DFNT_FLOAT32;
	fieldtype[2] = DFNT_CHAR8;
	
	fieldorder[0] = 1;
	fieldorder[1] = 4;
	fieldorder[2] = 4;
	
	status = PTdeflevel(PTid, "Sensor", fldlist, fieldtype, fieldorder);
	
	PTdetach(PTid);

	goto skip;

	/* Fixed Buoy Point */
	/* ---------------- */
	PTid = PTattach(ptfid, "FixedBuoy Point");


	/* Define Description/Location Level */
	strcpy(fldlist, "Label,Longitude,Latitude,DeployDate,ID");
	
	fieldtype[0] = DFNT_CHAR8;
	fieldtype[1] = DFNT_FLOAT64;
	fieldtype[2] = DFNT_FLOAT64;
	fieldtype[3] = DFNT_INT32;
	fieldtype[4] = DFNT_CHAR8;

	fieldorder[0] = 8;
	fieldorder[1] = 0;  /* Order 0 same as Order 1 for numeric scalars */
	fieldorder[2] = 0;
	fieldorder[3] = 0;
	fieldorder[4] = 1;

	status = PTdeflevel(PTid, "Desc-Loc", fldlist, 
			    fieldtype, fieldorder);


	/* Define Data Level */
	strcpy(fldlist, "Time,Rainfall,Temperature,ID");
	
	fieldtype[0] = DFNT_FLOAT64;
	fieldtype[1] = DFNT_FLOAT32;
	fieldtype[2] = DFNT_FLOAT32;
	fieldtype[3] = DFNT_CHAR8;

	fieldorder[0] = 0;
	fieldorder[1] = 0;
	fieldorder[2] = 0;
	fieldorder[3] = 1;

	status = PTdeflevel(PTid, "Observations", fldlist, 
			    fieldtype, fieldorder);

	printf("status from PTdeflevel: %d\n");
	
	status = PTdeflinkage(PTid, "Desc-Loc", "Observations", "ID");
	printf("status from PTdeflinkage: %d\n");
	status = PTdetach(PTid);
	printf("status from PTdetach: %d\n");



	

	/* Floating Buoy Point */
	/* ------------------- */
	PTid = PTattach(ptfid, "FloatBuoy Point");


	/* Define Description Level */
	strcpy(fldlist, "Label,DeployDate,Weight,ID");
	
	fieldtype[0] = DFNT_CHAR8;
	fieldtype[1] = DFNT_INT32;
	fieldtype[2] = DFNT_INT16;
	fieldtype[3] = DFNT_CHAR8;

	fieldorder[0] = 8;
	fieldorder[1] = 0;
	fieldorder[2] = 0;
	fieldorder[3] = 1;


	status = PTdeflevel(PTid, "Description", fldlist, 
			    fieldtype, fieldorder);


	/* Define Data Level */
	strcpy(fldlist, "Time,Longitude,Latitude,Rainfall,Temperature,ID");

	fieldtype[0] = DFNT_FLOAT64;
	fieldtype[1] = DFNT_FLOAT64;
	fieldtype[2] = DFNT_FLOAT64;
	fieldtype[3] = DFNT_FLOAT32;
	fieldtype[4] = DFNT_FLOAT32;
	fieldtype[5] = DFNT_CHAR8;

	fieldorder[0] = 0;
	fieldorder[1] = 0;
	fieldorder[2] = 0;
	fieldorder[3] = 0;
	fieldorder[4] = 0;
	fieldorder[5] = 1;	

	status = PTdeflevel(PTid, "Measurements", fldlist, 
			    fieldtype, fieldorder);

	status = PTdeflinkage(PTid, "Description", "Measurements", "ID");

	PTdetach(PTid);

      skip:
	PTclose(ptfid);
    }
    
    return;
}
